packages/apify/src/internals/charging.ts

```diff
@@ -0,0 +1,307 @@
+        }
+
+        return {
+            pricingModel: this.pricingModel,
+            isPayPerEvent: this.isPayPerEvent,
+            maxTotalChargeUsd: this.maxTotalChargeUsd,
+            perEventPrices: Object.fromEntries(
+                Object.entries(this.pricingInfo).map(([eventName, { price }]) => [eventName, price]),
+            ),
+        };
+    }
+
+    /**
+     * Charge for a specified number of events - sub-operations of the Actor.
+     *
+     * @param options The name of the event to charge for and the number of events to be charged.
+     */
+    async charge({ eventName, count = 1 }: ChargeOptions): Promise<ChargeResult> {
+        const calculateChargeableWithinLimit = () => Object.fromEntries(
+            Object.keys(this.pricingInfo).map(
+                (name) => [name, this.calculateMaxEventChargeCountWithinLimit(name)],
+            ),
+        );
+
+        if (!this.isPayPerEvent) {
+            if (!this.notPpeWarningPrinted) {
+                log.warning('Ignored attempt to charge for an event - the Actor does not use the pay-per-event pricing');
+                this.notPpeWarningPrinted = true;
+            }
+
+            return {
+                eventChargeLimitReached: false,
+                chargedCount: 0,
+                chargeableWithinLimit: calculateChargeableWithinLimit(),
+            };
+        }
+
+        if (this.chargingState === undefined) {
+            throw new Error('ChargingManager is not initialized');
+        }
+
+        /* START OF CRITICAL SECTION - no awaits here */
+        const chargedCount = Math.min(count, this.calculateMaxEventChargeCountWithinLimit(eventName));
+
+        if (chargedCount === 0) {
+            return {
+                eventChargeLimitReached: true,
+                chargedCount: 0,
+                chargeableWithinLimit: calculateChargeableWithinLimit(),
+            };
+        }
+
+        const pricingInfo = this.pricingInfo[eventName] ?? {
+            price: this.isAtHome ? 0 : 1, // Use a nonzero price for local development so that the maximum budget can be reached
+            title: `Unknown event '${eventName}'`,
+        };
+
+        this.chargingState[eventName] ??= { chargeCount: 0, totalChargedAmount: 0 };
+        this.chargingState[eventName].chargeCount += chargedCount;
+        this.chargingState[eventName].totalChargedAmount += chargedCount * pricingInfo.price;
+
+        /* END OF CRITICAL SECTION */
+
+        if (this.isAtHome) {
+            if (this.pricingInfo[eventName] !== undefined) {
+                await this.apifyClient.run(this.actorRunId!).charge({ eventName, count: chargedCount });
+            } else {
+                log.warning(`Attempting to charge for an unknown event '${eventName}'`);
+            }
+        }
+
+        const timestamp = new Date().toISOString();
+
+        if (this.chargingLogDataset !== undefined) {
+            for (let i = 0; i < chargedCount; i++) {
```

Comment:
Suggestion: why not push one item for the whole `chargedCount`? If the developer wants to simultaneously charge N things, I would expect a single item would suffice for debugging. Charging line by line will potentially create a lot of data with no real value added. Also, doing it like this rapidly increases the potential cost incurred by using this (by mistake) in production.

Reply:
This is taken from the PoC by @metalwarrior665, but I agree with your suggestion. It's small enough of a change to add that.

Reply:
The idea was that there will be mapping of items and events and if you push a list, each event might have different metadata. But we diverged from the idea anyway so feel free to change